xds/src/main/java/io/grpc/xds/XdsClientImpl.java

```diff
@@ -69,6 +69,21 @@
 final class XdsClientImpl extends XdsClient {
   private static final Logger logger = Logger.getLogger(XdsClientImpl.class.getName());
 
+  // Longest time to wait, since the subscription to some resource, for concluding its absence.
+  @VisibleForTesting
+  static final int INITIAL_RESOURCE_FETCH_TIMEOUT_SEC = 15;
+
+  // Reserved value for special usage, as a placeholder in CDS cache for cluster not exist.
+  private static final ClusterUpdate CLUSTER_NOT_FOUND =
+      ClusterUpdate.newBuilder()
+          .setClusterName("cluster not found")
```

Comment:
This is too hacky. Checking whether the resource is not found by checking whether the value of the resource key in a map is this special value is also too hacky. Maintaining a set for resources not found is more elegant.

Reply:
I was trying to maintain a set for resources not found. However, I found that approach is subtler than putting such a special placeholder in CDS/EDS caches. For example, if the CDS cache consists of two parts: a map of present resources and a set of absent resources, a cached resource can exist in either of them (if it exists, it is cached in the map; if not, it is cached in the set). This is harder to maintain. Every time a new watcher is added, we need to look up both of them. When the last watcher for a resource is removed, we also need to ensure the cached value no longer exists in either the map or the set. Something more cumbersome is that, when a new response comes such that a previously non-exist resource becomes present, we need to remove the entry from the set and add one into the map. Similar for some previously existing resource becomes absent. Therefore, I leverage the benefit of maintaining a single cache for resources that are present and absent.

Reply:
>Every time a new watcher is added, we need to look up both of them. 

The current approach need to check both the existence and whether it is the special value. Same complexity.

>When the last watcher for a resource is removed, we also need to ensure the cached value no longer exists in either the map or the set... 

This (and also to clean up the timers) is the right thing to do to clean up. I don't see any problem. 

>when a new response comes such that a previously non-exist resource becomes present, we need to remove the entry from the set and add one into the map.

This is the right operation to do. It's a hack to replace the special value (NOT_FOUND) with a valid value.

Reply:
Alright, changed.